text/0000-swap-out-jemalloc.md

```diff
@@ -0,0 +1,235 @@
+* If only an rlib is being produced, no allocator is injected.
+* If any crate tagged with `#[allocator]` has been explicitly linked to (e.g.
+  via an `extern crate` statement directly or transitively) then no allocator is
+  injected.
+* If two allocators have been linked to explicitly an error is generated.
+* If only a binary is being produced, then the target's `exe_allocation_crate`
+  value is injected, otherwise the `lib_allocation_crate` is injected.
+
+The compiler will also record that the injected crate is injected, so later
+compilations know that rlibs don't actually require the injected crate at
+runtime (allowing it to be overridden).
+
+### Allocators in practice
+
+Most libraries written in Rust wouldn't interact with the scheme proposed in
+this RFC at all as they wouldn't explicitly link with an allocator and generally
+are compiled as rlibs. If a Rust dynamic library is used as a dependency, then
+its original choice of allocator is propagated throughout the crate graph, but
+this rarely happens (except for the compiler itself, which will continue to use
+jemalloc).
+
+Authors of crates which are embedded into other runtimes will start using the
+system allocator by default with no extra annotation needed. If they wish to
+funnel Rust allocations to the same source as the host application's allocations
+then a crate can be written and linked in.
+
+Finally, providers of allocators will simply provide a crate to do so, and then
+applications and/or libraries can make explicit use of the allocator by
+depending on it as usual.
+
+# Drawbacks
+
+A significant amount of API surface area is being added to the compiler and
+standard distribution as part of this RFC, but it is possible for it to all
+enter as `#[unstable]`, so we can take our time stabilizing it and perhaps only
+stabilize a subset over time.
+
+The limitation of an allocator crate not being able to link to the standard
+library (or libcollections) may be a somewhat significant hit to the ergonomics
+of defining an allocator, but allocators are traditionally a very niche class of
+library and end up defining their own data structures regardless.
+
+Libraries on crates.io may accidentally link to an allocator and not actually
+use any specific API from it (other than the standard allocation symbols),
+forcing transitive dependants to silently use that allocator.
+
+This RFC does not specify the ability to swap out the allocator via the command
+line, which is certainly possible and sometimes more convenient than modifying
+the source itself.
+
+It's possible to define an allocator API (e.g. define the symbols) but then
+forget the `#![allocator]` annotation, causing the compiler to wind up linking
+two allocators, which may cause link errors that are difficult to debug.
+
+# Alternatives
+
+The compiler's knowledge about allocators could be simplified quite a bit to the
+point where a compiler flag is used to just turn injection on/off, and then it's
```

Comment:
I did not have much concrete in mind when I wrote this alternative, and I'm not quite sure what you're thinking of, so could you go into more detail about how you would envision "dumbing down" the compiler's knowlege of allocators? 

> I actually maintain a viewpoint that neither of these should have allocator built-in and producer of the final executable should link all the appropriate allocator libraries instead

Unfortunately a dynamic library will not link on Windows unless all symbols are resolved (unlike linux where you can have unresolved symbols in a dynamic library)